### ISSUE TITLE
Issues using SerialPort via Async(Read/Write)Ext

### ISSUE THREAD

**Issue by jangernert** [berkowski/tokio-serial#29]
I'm trying to port an existing C# application that talks to scales to rust. I went all in on async and tokio so I thought this crate would be a good fit. Currently only the windows build has been tested with a scale connected to a serial port.

It seems to me like the serial port is not really waiting for read/write/flush to finish.
Some example code still with debugging modifications:

```
async fn read_bytes(&self, number_of_bytes: usize) -> Result<Vec<u8>, Error> {
    let serial_port_mutex = self.port.as_ref().ok_or_else(|| {
        let msg = "Failed to read bytes: Serial Port not initialized";
        log::error!("{}", msg);
        Error::new(ErrorKind::NotConnected, msg)
    })?;
    log::info!("bytes to write: {:?}", serial_port_mutex.lock().await.bytes_to_write().unwrap());
    log::info!("bytes to read: {:?}", serial_port_mutex.lock().await.bytes_to_read().unwrap());
    let mut buffer = vec![0; number_of_bytes];
    serial_port_mutex.lock().await.read_exact(&mut buffer).await.map_err(|e| {
        log::warn!("Failed to read {} bytes: {}", number_of_bytes, e);
        e
    })?;
    log::info!("buffer: {:?}", buffer);
    Ok(buffer)
}

async fn write_bytes(&self, buffer: &[u8]) -> Result<(), Error> {
    let serial_port_mutex = self.port.as_ref().ok_or_else(|| {
        let msg = "Failed to write bytes: Serial Port not initialized";
        log::error!("{}", msg);
        Error::new(ErrorKind::NotConnected, msg)
    })?;
    serial_port_mutex.lock().await.write_all(buffer).await.map_err(|e| {
        log::error!("Failed to write bytes to serial port");
        e
    })?;
    serial_port_mutex.lock().await.flush().await.map_err(|e| {
        let msg = format!("Failed to flush serial port: {}", e);
        log::error!("{}", msg);
        Error::new(ErrorKind::BrokenPipe, msg)
    })?;
    // while serial_port_mutex.lock().await.bytes_to_write().unwrap() > 0 {
    //      tokio::time::delay_for(tokio::time::Duration::new(0, 20 * 1000000)).await;
    // }
    Ok(())
}
```

After writing a message to the scale, then flushing and trying to read the response of the scale the serial port indicates to still not have written the bytes from the out buffer to the port: `bytes to write: 15`.
The read then fails with: `Failed to read 1 bytes: The I/O operation has been aborted because of either a thread exit or an application request. (os error 995)`
Trying to force the program to wait until all bytes have been written with `delay_for` and then forcing it to wait again until there are bytes to read doesn't fix the problem either. Then the serial port indicates that there are no bytes to write and some bytes to read. But the read operation fails with the same error as before.

I know the only sample is using this framed de/encoder thing. But I'd rather directly write and read some bytes to the port.

**Comment by DanielJoyce** [comment]
First off the serial port is a lot slower than the code writing to it. It's possible the flush has written the bytes and you're trying to read before the scale has even responded yet. There is a lot less overhead in the rust side compared to C#.

The whole write / flush / read cycle is likely happening before the scale has even responded.

Also the behavior of serial ports is different on windows vs Linux.

In Linux a timeout on a serial port is how long to wait to try to read even a single byte.

On windows a serial timeout is how long to wait to before returning with data. So even if data is available windows will wait the full timeout to try and fill buffers then return. Of course this can be changed with settings. 

So the question is what OS are you on and what are the other settings.

**Comment by stefa168** [comment]
Reviving this issue as I am having the same issues on Linux Mint 21.2.
When doing repeated coil reads, the first returns no data, then the second has more data (all of the previous query, plus about half of the current read) and so on. We're missing the possibility to flush data not needed as well.

**Comment by 13thrule** [comment]
Confirming this still reproduces on current versions — tokio 1.53.1, tokio-serial 5.5.0, Windows 11.

Device: a Raspberry Pi Pico running MicroPython, presenting as a standard USB-CDC virtual COM port.

Symptom: `tokio_serial::new(port, baud).open_native_async()` returns `Ok` with no error, and `AsyncBufReadExt::lines()` / `.next_line().await` on the resulting stream never resolves — zero bytes ever observed, indefinitely. No panic, no timeout, nothing in the logs. The exact same port/device/baud read immediately and reliably via Python's `pyserial` (blocking) run right after killing the Rust process, so it's not a device, wiring, or driver issue.

Things I ruled out via controlled testing before concluding this is the async integration itself:
- DTR/RTS state — toggled both explicitly via pyserial (`dtr=False`/`rts=False` then `dtr=True`/`rts=True`); data flowed identically either way, so not gated on either signal.
- Explicit serial parameters — set `data_bits(Eight)`, `parity(None)`, `stop_bits(One)`, `flow_control(None)` explicitly on the builder to match pyserial's defaults exactly; no change.

Not a fix to this crate's async internals (the root cause looks like it needs a Windows equivalent of `AsyncFd` in Tokio itself, which is out of scope for a comment), but here's the workaround that resolved it for me — plain blocking `serialport` on a dedicated thread, bridged into async code via an mpsc channel, in case it saves someone else the time it cost me:

```rust
let (line_tx, mut line_rx) = tokio::sync::mpsc::unbounded_channel::<String>();
let port_owned = port.to_string();

std::thread::spawn(move || {
    let serial = serialport::new(&port_owned, baud)
        .timeout(std::time::Duration::from_millis(500))
        .open()
        .expect("failed to open port");
    let mut reader = std::io::BufReader::new(serial);
    loop {
        let mut line = String::new();
        match std::io::BufRead::read_line(&mut reader, &mut line) {
            Ok(0) => break, // port closed
            Ok(_) => {
                if line_tx.send(line.trim().to_string()).is_err() {
                    break; // receiver dropped
                }
            }
            Err(e) if e.kind() == std::io::ErrorKind::TimedOut => continue,
            Err(_) => break,
        }
    }
});

// async side
while let Some(line) = line_rx.recv().await {
    // handle line
}
```

Reliable, no further issues since switching to this.

Happy to share more repro detail if it'd help — wanted to leave a data point here first since this matches the reported symptom closely and I lost a fair bit of time assuming it was a hardware/wiring problem before finding this issue.
